core: Correct and implement MovieClip states (default state, initial loading state, error state, image state and unloaded state)

Korne127 opened the pull request:
**Edit: The scope of this pull request has drastically increased. To read the current summary of it, [click here](https://github.com/ruffle-rs/ruffle/pull/10792#issuecomment-1612309656) (or scroll down).**

----



If an SWF file tries to load a movie that can't be loaded (e.g. because it doesn't exist), the Flash Player sets _framesloaded of the MovieClip to -1.
This has previously not been implemented in Ruffle (which returned 0).

This caused [Paraplüsch](http://parapluesch.de/whiskystore/test.htm) / Paraplush to not work in Ruffle and being stuck in the menu. It need loads an additional SWF if you click on a patient and tries two links (an alternative link if the first one doesn't work). Since the original links don't exist anymore and can't be accessed because of CORS, it needs to load the working alternative link, but the error handling code has previously not been invoked due to the false Ruffle behaviour.

This pull request fixes this issue and sets _framesloaded of the MovieClip to -1 if the movie couldn't be loaded.
It also removes some duplicated functions in the code.


~~However, while testing it, I've noticed a follow-up issue. In Paraplüsch, the movie is loaded via `loadMovie(path, target);`. (The code responsible for this loading is in `scripts/DefineSprite (1862)/frame 1/DoAction`.)
But if you replace that function with `loadMovie(path, target, "GET");`, loading the patient doesn't work in Ruffle anymore, although it still worked with the flash plugin. (I just noticed this because FFDec automatically does this when recompiling the file.)
I tried debugging this, but I don't really know what caused this. It would be good if someone who is familiar with the loadMovie implementation looks over this sometimes. I'll probably open this as an issue if this PR gets merged.~~
Edit: I noticed this was just an FFDec bug; recompiling the file changed the bytecode and made it not work in both Ruffle and Flash itself (I must have tested the wrong file back when I wrote this). This was unrelated to the "GET".


If someone think that this PR is okay and can be merged, please just tell me and I'll rebase it onto the latest master branch so you can merge it (instead of rebasing using the Github feature). :)

Dinnerbone commented:
Thank you! Would it be possible for you to add a test for this?

Lord-McSweeney commented:
Is this also the correct behavior for AVM2?

Korne127 commented:
Thanks for the quick response; I'll try writing some tests and testing it in AVM2 as well :)

Korne127 commented:
@Lord-McSweeney Okay, I looked a bit into it and there is no possibility to access MovieClip variables in ActionScript 3 if the clip couldn't be loaded.
You use a `Loader` to load the SWF and usually use `loader.content as MovieClip` to get the MovieClip and access its variables if it has been loaded (also possible with e.g. `event.target.loader.content` in the onLoadComplete function). However, this is not possible in the error callback function (since `loader.content` is `null`), so you can't access the variables like `framesLoaded`. Therefore, the return value of it doesn't matter in ActionScript 3 at all if the movie couldn't be loaded.

Korne127 commented:
@Dinnerbone I've spent some time writing detailed tests which test the default variables and the changes of the MovieClip state (this made me notice that not only _framesloaded but also other variables change in this situation).

However, when trying to integrate them in Ruffle, I'm currently facing the problem that the _url property is also tested. When I execute that test, it includes my project root (`file:///<MyRoot>/ruffle/tests/tests/swfs/…`). But I need to adapt that test in a way that it passes on Ruffle. Therefore, I need to know what the file path is if this is executed via the Github CI so that the test passes on Github.

Toad06 commented:
I'd say to split the URL to only retain the common parts. This is what I did in a past PR:

```actionscript
var url = unescape(mc._url).split("/");
url = url[url.length - 2] + "/" + url[url.length - 1];
trace("mc._url (unescaped & split): " + url);
```

Traces:
> mc._url (unescaped & split): movieclip_methods_with_loaded_image/target.png

Korne127 commented:
Thank you; that's a really good solution!

Korne127 commented:
I've converted this to a draft because I've noticed some other deviations to the Flash player during my tests; I'll reopen this when I've finished the tests and implemented the changes in Ruffle.

Aaron1011 commented on core/src/display_object/movie_clip.rs:
```diff
@@ -2406,6 +2408,72 @@ impl<'gc> MovieClip<'gc> {
+    // calls with one frame delay? Does transform_to_unloaded_state need to get executed
+    // after one frame if the target is a MovieClip? Test the behaviour and adapt the code
+    // if necessary.
+    pub fn unload_movie(&self, context: &mut UpdateContext<'_, 'gc>) {
```
Can you rename this to `avm1_unload_movie`?

Korne127 replied:
Hey, sorry for the late response, I just wanted to finish the PR first.
But thanks for the suggestion; I can do that! :)

There's also MovieClip::avm1_unload (which is also mentioned in the documentation of this method); do you think there should be further differenciation between these methods if their names are so similar or is it okay like this?

Korne127 commented:
I finally finished implementing all the MovieClip states. Since the scope of this pull request has increased so much, this is a new summary of all of its changes:

### General concept:
A MovieClip can be in different specific states. The state of a MovieClip consists of the values of all properties and the results of some getter functions of the MovieClip. Previously, the states have not been correctly implemented, which can lead to flash content not properly working (e.g. Paraplüsch / Paraplush didn't recognise the error state because Ruffle didn't implement it correctly; therefore it didn't work).
This pull request implements all MovieClip states correctly and therefore fixes such bugs.

### New MovieClip states that have been implemented:
- **Error state**
The MovieClip now enters the error state if no file could be loaded or if the loaded content is no valid supported content. In the error state, some attributes have certain error values to signal that no valid file could be loaded.
- **Initial loading state**
The MovieClip now enters the initial loading state if it's loaded. In the initial loading state, some attributes have certain initial loading values to signal that the file is currently being loaded and neither an error has occurred nor the first frame has been successfully loaded yet.
- **Unloaded state**
The MovieClip now enters the unloaded state one frame after a command to unload the MovieClip has been read. In the unloaded state, some attributes have certain values.

### MovieClip states that have been corrected:
- **Image state**
The MovieClip enters the image state if an image has been loaded. Previously, some values in the image state haven't been set correctly. This has been fixed.
- **Default state**
The MovieClip enters the default state when it's created with `createEmptyMovieClip`. Previously, some values in the default state haven't been set correctly. This has been fixed.

### Added tests:
- **MovieClip state tests**
To make sure that all states are implemented correctly and will continue to do so, two tests testing the different states of a MovieClip have been added. `movieclip_default_state` tests the default state of a MovieClip and `movieclip_state_values` has MovieClips going through all other states and tests if the values are correct in each.
These tests currently only test the MovieClip states when trying to load a local file since Ruffle tests currently don't support tests using remote URLs. Test cases for remote URLs are still included and can be used if Ruffle starts supporting remote URLs at some point (e.g. with a local server).
- **MovieClip invalid getBounds tests**
The behaviour of when the results of `getBounds` / `getRect` calls on MovieClips with invalid bounds are `6710886.35` and when they are `6710886.4` for each corner of the rectangle is rather complex. To make sure it works correctly in every case and will continue to do so, eight tests testing the results of `getBounds` calls under these circumstances have been added.
This needs to be tested in several tests because an internal state determining the results can change irreversibly. Making sure that it changes correctly on different occasions takes several tests.

### Other big changes:
- All `NavigatorBackend` implementations have been refactored, resulting in improved code quality, less duplicated code, more consistent and easier to understand procedures, additional error handling and better error messages.
- The calculation of the correct results of `getBounds` / `getRect` calls on MovieClips with invalid bounds has been implemented.

### Minor changes:
- Documentation has been added and improved.
- Duplicated functions have been removed.
- Some TODOs about other (possible) differences between Flash and Ruffle implementations, that are out of scope for this pull request, have been added.
- A spelling mistake has been corrected.

To implement this all, various functions, variables and structs have been created and changed. The individual commit messages contain more detailed information about that.

Korne127 commented:
Currently, there are just two minor problems. These are also the reason why some of the tests fail:
The `movieclip_default_state` test tests the values of all properties and the results of some getter functions after it has been created with `createEmptyMovieClip`.

However, due to a regression, AVM1 code currently can’t access the `filters`. Therefore, this test fails since it also tests the `filters` (which are part of the state) but can’t access them. This will hopefully be fixed soon.

Additionally to that, the `getTextSnapshot()` MovieClip method (and the `TextSnapshot` object) are currently not implemented or stubbed in Ruffle. Therefore, `getTextSnapshot().getCount()` just returns `undefined` in Ruffle and not the expected `0`, which it should in the default state.
I have tried stubbing out the `TextSnapshot` object for some time, but creating a new Avm-`Object` is part of a pretty different area in Ruffle in which I don’t really have experience in. I still tried it, but I couldn’t easily figure out how to properly create such an Avm-`Object` and connect it to the Rust struct.
Therefore, I’d really appreciate it if someone else, who is more experienced in that area, could help and just stub it out so that it returns the default values and Ruffle passes the test correctly.

Lord-McSweeney commented on tests/tests/swfs/avm1/movieclip_default_state/test.as:
```diff
@@ -0,0 +1,83 @@
+	mcProps.push(["getRect(mc).yMin", getRectMc.yMin]);
+	mcProps.push(["getRect(mc).yMax", getRectMc.yMax]);
+	mcProps.push(["getSWFVersion()", mc.getSWFVersion()]);
+	mcProps.push(["getTextSnapshot().getCount()", mc.getTextSnapshot().getCount()]);
```
You could probably just comment out this line, as well as the filters line, until they get imlemented, and add a TODO notice.

Korne127 replied:
Tbh yeah, it probably makes most sense to do that for now. I just hope I remember to add it back again if they get implemented.

Lord-McSweeney commented on scanner/src/file_results.rs:
```diff
@@ -85,7 +85,7 @@ pub struct FileResults {
     /// The uncompressed length of the SWF file.
     #[serde(rename = "Uncompressed Length")]
-    pub uncompressed_len: Option<u32>,
+    pub uncompressed_len: Option<i32>,
```
This doesn't really make much sense here (or possibly in some other places) - would it be good to leave it as a `u32` and do relevant coercions in places where they're needed?

Korne127 replied:
`uncompressed_len` needs to be an `i32` because in the error state `getBytesTotal()` returns `-1`. And `getBytesTotal()` returns `uncompressed_len`. Because there also exist valid states in which `getBytesTotal()` returns `0`, we can't just return `-1` if uncompressed_len is `0`.

To implement the error state correctly as Flash player does it, `uncompressed_len? needs to be `i32` so it can be set to `-1` in the error case.

Korne127 commented:
@Aaron1011 @Lord-McSweeney Do you have other comments about this pull request or my answers? If not; I'd add a commit adapting it to the reviews.

Aaron1011 commented on core/src/avm1/globals/movie_clip.rs:
```diff
@@ -1418,7 +1431,23 @@ fn get_bounds<'gc>(
+
+            // If the bounds are invalid, the target space is identical to the origin space and
+            // use_new_invalid_bounds_value is true, the returned bounds use a specific invalid value.
+            if *activation.context.use_new_invalid_bounds_value
```
Does this really need to be a field in `UpdateContext`? Could you instead perform the check

```rust
            if activation.swf_version() >= 8
                || movie_clip.get_oldest_parent_swf_version(&mut activation.context) >= 8
```

after you've found that `bounds` and `target_bounds` are invalid?

Korne127 replied:
Sadly, it needs to be. The behaviour in Flash is a bit weird: As soon as the call returned `X.4` for one time, it will always return `X.4` for invalid bounds with a different target. It is an actual internal state changing irreversibly.

So e.g. there can be the following scenario (that's also tested in the tests):
- Parent MovieClip with SWF version 7 calls `getBounds` => the check is `false` and it returns `X.35`
- Then, child movie clip with SWF version 8 is loaded and calls `getBounds` => the check is `true` and it returns `X.4`
- Parent MovieClip then calls `getBounds` again => even if the check is still `false`, it returns `X.4` (because the state changed with the second call).

Because this behaviour is so specific and complex, I added quite some `movieclip_invalid_bounds` tests making sure Ruffle exactly behaves like this.

kmeisthax commented on swf/src/types.rs:
```diff
@@ -97,7 +97,7 @@ pub struct HeaderExt {
     pub(crate) file_attributes: FileAttributes,
     pub(crate) background_color: Option<SetBackgroundColor>,
-    pub(crate) uncompressed_len: u32,
+    pub(crate) uncompressed_len: i32,
```
I feel a little weird letting negative lengths leak into the SWF parsing machinery. Did Flash have a hard 2GB cap on SWF size (even on 64-bit)?

Korne127 replied:
I understand that. Sadly we can't just use an unsigned integer here because `-1` is a valid value needed in some situations (and `0` is as well, so we can't just read `0` as `-1`).
One alternative to this I easily see would be an `Option<u32>` (with `None` standing for `-1`). However, this would make comparisons to other integers much harder since these multiple cases would need to be handled in each comparison (or a helper function would need to be used, which also isn't that clean).

The other possible alternative to this I see would be a custom `enum` with `u32` and a dedicated `MinusOne` value. This has the advantage that `PartialEq<u32>` and `PartialOrd<u32>` could be implemented for it (and vice versa), so that all comparisons could just work without changing them in any way. This might make sense in general because there are other variables in Ruffle that can just be positive or `-1` as well (e.g. frames_loaded which this pull request has also changed (from `u16` to `i32`)).

What do you think about those ideas?

And no, at least I'm not aware of any hard SWF file size cap. I just used `i32` because it has previously been `u32` and the 2GB are so large that it seemed to make more sense to change it to `u32` than `u64`. The same thought applies to a 4GB file with all the `u32`s used to store the SWF file size though.

kmeisthax replied:
Would it be possible to store the "this is actually an error SWF" flag in `SwfMovie`, and then have `SwfMovie.uncompressed_len()` check for that and return `-1` as necessary? (And, of course, not access `SwfMovie.header().uncompressed_len` directly)

I'm less concerned about the file size cap and more with letting details of how Ruffle works determine `swf`'s interface. If you're doing low-level SWF stuff with `swf` you only care about the uncompressed file size and not any error conditions that might have been signaled by Flash Player abusing that field.

Korne127 replied:
That might be a possibility as well. Although this error condition can only appear if there literally is no existing SWF file, so in any case where it would be `-1`, there is no actual file size you could care about.

Still, if you prefer it, I can change it to this. The only thing worrying me about this is that the identically named functions `SwfMovie::uncompressed_len` and `HeaderExt::uncompressed_len` would not always return the same value (although this would be mitigated by documentation). And there are some places where `HeaderExt::uncompressed_len` is directly used: To create `MovieMetadata` in `web/src/lib.rs` and to set the `FileResults` in `scanner/src/execute.rs`. I don't know if these values are furtherly used elsewhere and problems arise if these values used directly from the header diverge from the `-1` error values retrieved by the `MovieClip`.
I could change these places to use the `SwfMovie::uncompressed_len` value, but I'm not sure if it would be good to do that or if there are reasons why they take it from the header directly and maybe should not be `-1` in such an error situation.
What would you think about that all?

Korne127 replied:
I've played around with this for a while and reread your previous statement, and I noticed your "(And, of course, not access `SwfMovie.header().uncompressed_len` directly)", which I think would probably mean that I should change `web/src/lib.rs` and `scanner/src/execute.rs` to use `SwfMovie::uncompressed_len` instead.
Since my biggest worry was that there were multiple variables containing different values with the same name (`uncompressed_len`) and that both of these are used in many cases, resulting in two diverging values being used, this would be mitigated by just using `SwfMovie::uncompressed_len` everywhere (and adding documentation stating that only it should be used).

I still don't really understand what speaks against just directly setting `HeaderExt::uncompressed_len` to the real Flash value. I personally think that using something like an `U32OrMinusOne` enum would be a good solution.
In this error case, no SWF file and no "correct" value exists (as there is no file size); the whole header is then just consisting of the error values Flash uses, so in my opinion, having a dedicated error value would be coherent and made sense in the header. An `U32OrMinusOne` enum allows the correct range of possible values.
And by using this, we could be sure `uncompressed_len` is always the same and that there aren't diverging variables with the same name, no matter which function is used (or will be added in the future) to get it, so for me, this look a cleaner solution.

However, these are just my thoughts about it. If there are other reasons why you prefer it with a flag, that's fine with me as well. We could also still use such an enum as the `HeaderExt::uncompressed_len` return value; that would solve the 2GB cap and be a clean way to allow `-1` but not other illegal negative values. The only downside to that enum I see is that it needs 8 bytes (as much as an `i64`).

kmeisthax replied:
Y'know, I think I might be overthinking this. Let's just keep this as-is for now. If it causes a problem later on we can fix it.

kmeisthax commented on tests/tests/swfs/avm1/movieclip_state_values/test.as:
```diff
@@ -0,0 +1,609 @@
+ * The following code is unused since Ruffle tests currently don't support tests using remote URLs.
+ * It is kept here because Flash behaves differently with online SWFs and offline SWFs.
+ * If Ruffle tests start supporting remote URLs (e.g. with a local server) at some point, these tests
+ * can still be used.
```
We probably could change the test harness to redirect remote URLs to specific files or simulate specific network conditions. Definitely out of scope for this PR...

Korne127 replied:
Yeah, Adrian17 said that it would be possible to use a local server or some kind of redirection to simulate tests with remote files.
So if this is done after this PR is merged, we can easily start using those tests as well in another PR. :)

kmeisthax reviewed (approved):
Excellent PR! There's numerous tests here and it's clear you spent a lot of time getting to the bottom of this.

Most of the comments I noted here aren't merge-stoppers.

Korne127 commented on core/src/backend/navigator.rs:
```diff
@@ -204,7 +214,14 @@ pub trait NavigatorBackend {
+    ///
+    /// If the URL is local, this equals the URL returned by fetch. Otherwise,
+    /// fetch may return a different URL, e.g. considering redirections.
+    fn resolve_url(&self, url: &str) -> Result<Url, ParseError>;
```
Yes, `fetch` takes a `Request` struct with an URL string (and returns a response with the redirected `Url`). The `fetch` implementations now use `resolve_url` internally to create an `Url` out of the string and do the necessary preprocessing.

kmeisthax replied:
Gotcha.

Korne127 commented on core/src/loader.rs:
```diff
@@ -1916,6 +1956,76 @@ impl<'gc> Loader<'gc> {
+    ) {
+        // If a local URL is fetched using the flash plugin, the _url property
+        // won't be changed => It keeps being the parent SWF URL.
+        if cfg!(target_family = "wasm") {
```
How would you suggest that? I could imagine a `NavigatorBackend` method like `isOnWeb` which returns true in the `WebNavigatorBackend` and false in the others. If the wasm check isn't clean to test whether we're in a browser, this might be a better idea (although a cfg like this is used in two other places in Ruffle; it might make sense to replace them as well).

The test itself should be done in this method; otherwise it would need to be implemented in different places for both the initial loading and the error state since this applies to both and the method is called for both (if the file is local).

Lord-McSweeney replied:
Checking the compile architecture to determine the build type is something that we do in many places.

kmeisthax replied:
We might want to have an explicit feature for at least some of the things that currently rely on `target_family='wasm'`. Again, entirely out of scope. For this particular PR alone I was more confused as to why `NavigatorBackend`'s fancy new `parse_url` method wasn't being used to change `file://` URLs.

Korne127 replied:
The new `resolve_url` method is just used to actually resolve the url string and do the necessary preprocessing, which depends on the `NavigatorBackend` implementation. It is used by both the `navigate_to_url` (which opens a URL link) and the `fetch` (which loads a movie) methods.
These methods need to know the correct resolved URL. E.g. if `navigate_to_url` is called with a local URL string, it needs to know that it's local, throws an error and doesn't open it.
What this code does, is setting the `SwfMovie::url` variable. And if ActionScript code tries to load a local file on web, Flash Player makes the MovieClip go into the error state but doesn't change the url, FP "lies about it" and it just stays the same.

But this just works if `resolve_url` returns the correct `Url` and this condition is specifically tested afterwards. If `resolve_url` itself would return this in the situation, `fetch` and `navigate_to_url` would think it's a valid url (the url of the parent SWF itself) and load / navigate there. So yeah, they need to know it's a local file / invalid and `resolve_url` needs to actually just resolve the URL correctly, but because FP lies in the error state, when setting this state, this needs to be deliberately tested and taken into account to set the state as FP does.

kmeisthax replied:
Gotcha.

n0samu replied:
The fetch API doesn't allow access to `file` URLs, so all of this is really inconsequential/pointless, right?

Korne127 replied:
@n0samu What exactly do you mean? The code highlighted in this review sets the `url` property of the `MovieClip, which is accessible in ActionScript. If this isn't done, the `url` property of a `MovieClip` is wrong under these specific circumstances.
In general, the `NavigationBackend` refactor allows for better code quality and improves the error messages. Previously, you could also not load local files on web, but a generic "Got JS error" got returned, while now that case is directly handled when resolving the URL and a precise error message is returned.

n0samu replied:
Oh okay, thanks for the explanation, don't mind me

Korne127 replied:
Dw; it's a good question
And you're welcome :)

Korne127 commented on swf/src/types.rs:
```diff
@@ -112,8 +112,19 @@ impl HeaderExt {
 
+    /// Returns the header for the error state movie stub which is used if no file
+    /// could be loaded or if the loaded content is no valid supported content.
+    pub fn default_error_header() -> Self {
```
I thought it would be cleaner to have the different default headers that Ruffle uses for an SWF in different states all in the same place. Especially since both the success header (which is used in the success state) and the image header (which is used in the image state) are located here, I think the same makes sense for the error header (used in the error state).

kmeisthax replied:
Aah, in that case we'll have to keep with what's already there. Maybe a future PR could move that out of `swf`.

Korne127 commented:
I rebased this pull request now and resolved all conflicts. While doing that, I removed the change commenting out the filters test as they're now implemented in AVM1.
